// ngx-img-cropper/esm5/lib/image-cropper/imageCropperDataShare.js
var ImageCropperDataShare = /** @class */ (function () {
    function ImageCropperDataShare() {
        this.share = {};
    }
    ImageCropperDataShare.prototype.setPressed = function (canvas) {
        this.pressed = canvas;
    };
    ImageCropperDataShare.prototype.setReleased = function (canvas) {
        if (canvas === this.pressed) {
            //  this.pressed = undefined;
        }
    };
    ImageCropperDataShare.prototype.setOver = function (canvas) {
        this.over = canvas;
    };
    ImageCropperDataShare.prototype.setStyle = function (canvas, style) {
        if (this.pressed !== undefined) {
            if (this.pressed === canvas) {
                // TODO: check this
                // angular.element(document.documentElement).css('cursor', style);
            }
        }
        else {
            if (canvas === this.over) {
                // TODO: check this
                // angular.element(document.documentElement).css('cursor', style);
            }
        }
    };
    return ImageCropperDataShare;
}());
export { ImageCropperDataShare };
