// ng2-validation/__ivy_ngcc__/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var base64_1 = require("./base64");
var credit_card_1 = require("./credit-card");
var date_1 = require("./date");
var date_ios_1 = require("./date-ios");
var digits_1 = require("./digits");
var email_1 = require("./email");
var equal_1 = require("./equal");
var equal_to_1 = require("./equal-to");
var greater_than_1 = require("./greater-than");
var greater_than_equal_1 = require("./greater-than-equal");
var json_1 = require("./json");
var less_than_1 = require("./less-than");
var less_than_equal_1 = require("./less-than-equal");
var max_1 = require("./max");
var max_date_1 = require("./max-date");
var min_1 = require("./min");
var min_date_1 = require("./min-date");
var not_equal_1 = require("./not-equal");
var not_equal_to_1 = require("./not-equal-to");
var number_1 = require("./number");
var phone_1 = require("./phone");
var range_1 = require("./range");
var range_length_1 = require("./range-length");
var url_1 = require("./url");
var uuid_1 = require("./uuid");
exports.CustomValidators = {
    base64: base64_1.base64,
    creditCard: credit_card_1.creditCard,
    date: date_1.date,
    dateISO: date_ios_1.dateISO,
    digits: digits_1.digits,
    email: email_1.email,
    equal: equal_1.equal,
    equalTo: equal_to_1.equalTo,
    gt: greater_than_1.gt,
    gte: greater_than_equal_1.gte,
    json: json_1.json,
    lt: less_than_1.lt,
    lte: less_than_equal_1.lte,
    max: max_1.max,
    maxDate: max_date_1.maxDate,
    min: min_1.min,
    minDate: min_date_1.minDate,
    notEqual: not_equal_1.notEqual,
    notEqualTo: not_equal_to_1.notEqualTo,
    number: number_1.number,
    phone: phone_1.phone,
    range: range_1.range,
    rangeLength: range_length_1.rangeLength,
    url: url_1.url,
    uuid: uuid_1.uuid
};
var ɵngcc0 = require('@angular/core');
var ɵngcc1 = require('./base64/directive');
var ɵngcc2 = require('./credit-card/directive');
var ɵngcc3 = require('./date/directive');
var ɵngcc4 = require('./date-ios/directive');
var ɵngcc5 = require('./digits/directive');
var ɵngcc6 = require('./email/directive');
var ɵngcc7 = require('./equal/directive');
var ɵngcc8 = require('./equal-to/directive');
var ɵngcc9 = require('./greater-than/directive');
var ɵngcc10 = require('./greater-than-equal/directive');
var ɵngcc11 = require('./json/directive');
var ɵngcc12 = require('./less-than/directive');
var ɵngcc13 = require('./less-than-equal/directive');
var ɵngcc14 = require('./max/directive');
var ɵngcc15 = require('./max-date/directive');
var ɵngcc16 = require('./min/directive');
var ɵngcc17 = require('./min-date/directive');
var ɵngcc18 = require('./not-equal/directive');
var ɵngcc19 = require('./not-equal-to/directive');
var ɵngcc20 = require('./number/directive');
var ɵngcc21 = require('./phone/directive');
var ɵngcc22 = require('./range/directive');
var ɵngcc23 = require('./range-length/directive');
var ɵngcc24 = require('./url/directive');
var ɵngcc25 = require('./uuid/directive');
var CUSTOM_FORM_DIRECTIVES = [
    base64_1.Base64Validator,
    credit_card_1.CreditCardValidator,
    date_1.DateValidator,
    date_ios_1.DateISOValidator,
    digits_1.DigitsValidator,
    email_1.EmailValidator,
    equal_1.EqualValidator,
    equal_to_1.EqualToValidator,
    greater_than_1.GreaterThanValidator,
    greater_than_equal_1.GreaterThanEqualValidator,
    json_1.JSONValidator,
    less_than_1.LessThanValidator,
    less_than_equal_1.LessThanEqualValidator,
    max_1.MaxValidator,
    max_date_1.MaxDateValidator,
    min_1.MinValidator,
    min_date_1.MinDateValidator,
    not_equal_1.NotEqualValidator,
    not_equal_to_1.NotEqualToValidator,
    number_1.NumberValidator,
    phone_1.PhoneValidator,
    range_1.RangeValidator,
    range_length_1.RangeLengthValidator,
    url_1.UrlValidator,
    uuid_1.UUIDValidator
];
var CustomFormsModule = (function () {
    function CustomFormsModule() {
    }
CustomFormsModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: CustomFormsModule });
CustomFormsModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function CustomFormsModule_Factory(t) { return new (t || CustomFormsModule)(); } });
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(CustomFormsModule, { declarations: [ɵngcc1.Base64Validator, ɵngcc2.CreditCardValidator, ɵngcc3.DateValidator, ɵngcc4.DateISOValidator, ɵngcc5.DigitsValidator, ɵngcc6.EmailValidator, ɵngcc7.EqualValidator, ɵngcc8.EqualToValidator, ɵngcc9.GreaterThanValidator, ɵngcc10.GreaterThanEqualValidator, ɵngcc11.JSONValidator, ɵngcc12.LessThanValidator, ɵngcc13.LessThanEqualValidator, ɵngcc14.MaxValidator, ɵngcc15.MaxDateValidator, ɵngcc16.MinValidator, ɵngcc17.MinDateValidator, ɵngcc18.NotEqualValidator, ɵngcc19.NotEqualToValidator, ɵngcc20.NumberValidator, ɵngcc21.PhoneValidator, ɵngcc22.RangeValidator, ɵngcc23.RangeLengthValidator, ɵngcc24.UrlValidator, ɵngcc25.UUIDValidator], exports: [ɵngcc1.Base64Validator, ɵngcc2.CreditCardValidator, ɵngcc3.DateValidator, ɵngcc4.DateISOValidator, ɵngcc5.DigitsValidator, ɵngcc6.EmailValidator, ɵngcc7.EqualValidator, ɵngcc8.EqualToValidator, ɵngcc9.GreaterThanValidator, ɵngcc10.GreaterThanEqualValidator, ɵngcc11.JSONValidator, ɵngcc12.LessThanValidator, ɵngcc13.LessThanEqualValidator, ɵngcc14.MaxValidator, ɵngcc15.MaxDateValidator, ɵngcc16.MinValidator, ɵngcc17.MinDateValidator, ɵngcc18.NotEqualValidator, ɵngcc19.NotEqualToValidator, ɵngcc20.NumberValidator, ɵngcc21.PhoneValidator, ɵngcc22.RangeValidator, ɵngcc23.RangeLengthValidator, ɵngcc24.UrlValidator, ɵngcc25.UUIDValidator] }); })();
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CustomFormsModule, [{
        type: core_1.NgModule,
        args: [{
                declarations: [CUSTOM_FORM_DIRECTIVES],
                exports: [CUSTOM_FORM_DIRECTIVES]
            }]
    }], function () { return []; }, null); })();
    return CustomFormsModule;
}());
/** @nocollapse */
CustomFormsModule.ctorParameters = function () { return []; };
exports.CustomFormsModule = CustomFormsModule;

exports.Base64Validator = ɵngcc1.Base64Validator;
exports.CreditCardValidator = ɵngcc2.CreditCardValidator;
exports.DateValidator = ɵngcc3.DateValidator;
exports.DateISOValidator = ɵngcc4.DateISOValidator;
exports.DigitsValidator = ɵngcc5.DigitsValidator;
exports.EmailValidator = ɵngcc6.EmailValidator;
exports.EqualValidator = ɵngcc7.EqualValidator;
exports.EqualToValidator = ɵngcc8.EqualToValidator;
exports.GreaterThanValidator = ɵngcc9.GreaterThanValidator;
exports.GreaterThanEqualValidator = ɵngcc10.GreaterThanEqualValidator;
exports.JSONValidator = ɵngcc11.JSONValidator;
exports.LessThanValidator = ɵngcc12.LessThanValidator;
exports.LessThanEqualValidator = ɵngcc13.LessThanEqualValidator;
exports.MaxValidator = ɵngcc14.MaxValidator;
exports.MaxDateValidator = ɵngcc15.MaxDateValidator;
exports.MinValidator = ɵngcc16.MinValidator;
exports.MinDateValidator = ɵngcc17.MinDateValidator;
exports.NotEqualValidator = ɵngcc18.NotEqualValidator;
exports.NotEqualToValidator = ɵngcc19.NotEqualToValidator;
exports.NumberValidator = ɵngcc20.NumberValidator;
exports.PhoneValidator = ɵngcc21.PhoneValidator;
exports.RangeValidator = ɵngcc22.RangeValidator;
exports.RangeLengthValidator = ɵngcc23.RangeLengthValidator;
exports.UrlValidator = ɵngcc24.UrlValidator;
exports.UUIDValidator = ɵngcc25.UUIDValidator;
